### PR TITLE
Implement base for battle AI

### PR DESCRIPTION
**Pull request by idshibanov**
Skeleton unit logic only, still missing castle defend and combat spell heuristics.
New pathfinding algorithm for battle arena and special debug mode for it if DBG_DEVEL flag is enabled.

### CODE REVIEW

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 1]
```diff
@@ -34,30 +34,31 @@ using namespace Battle;
...
 {
-    bool CheckBattleRetreat()
+    bool CheckBattleRetreat( double myArmy, double enemy )
     {
-        // FIXME: force AI retreat from battle since battle logic is not implemented
-        return true;
+        // FIXME: more sophisticated logic to see if remaining units are under threat
+        // Consider taking speed/turn order into account as well
+        // Pass in ( const Units & friendly, const Units & enemies ) instead
+        return myArmy * 10 < enemy;
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:42]: It's good to know what **10** stands for.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 2]
```diff
@@ -34,30 +34,31 @@ using namespace Battle;
...
-        if ( CheckBattleRetreat() ) {
-            // Cast maximum damage spell
-
-            actions.push_back( Command( MSG_BATTLE_RETREAT ) );
-            actions.push_back( Command( MSG_BATTLE_END_TURN, currentUnit.GetUID() ) );
-            return;
-        }
+        if ( currentUnit.isFlying() == target.isFlying() )
+            return currentUnit.GetSpeed() > target.GetSpeed();
+        return currentUnit.isFlying();
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:54]: What if we have a walking unit but faster than current one?

File: src/fheroes2/battle/battle_pathfinding.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    int GetValidMoveIndex( int fromCell, int directionMask, bool isWide )
+    {
+        int newIndex = -1;
+        int wideUnitOffset = 0;
+        const int x = fromCell % ARENAW;
+        const int y = fromCell / ARENAW;
+
+        switch ( directionMask ) {
+        case Battle::TOP_LEFT:
+            if ( y > 0 && ( x > 0 || ( y % 2 ) == 0 ) ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_pathfinding.cpp:35]: We could cache `y % 2` like `const int mod = y % 2;`

File: src/fheroes2/battle/battle_pathfinding.cpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    ArenaPathfinder::ArenaPathfinder()
+        : _cache( ARENASIZE )
+    {}
+
+    void ArenaPathfinder::reset()
+    {
+        for ( size_t i = 0; i < _cache.size(); ++i ) {
+            _cache[i]._from = -1;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_pathfinding.cpp:87]: Let's cache `_cache[i]` by reference variable.

File: src/fheroes2/battle/battle_troop.cpp [hunk 5]
```diff
@@ -474,6 +474,16 @@ u32 Battle::Unit::GetSpeed( bool skip_standing_check ) const
...
 }
 
+uint32_t Battle::Unit::CalculateRetaliationDamage( uint32_t damageTaken ) const
+{
+    // Check if there will be retaliation in the first place
+    if ( damageTaken > hp || !AllowResponse() )
+        return 0;
+
+    const uint32_t unitsLeft = ( hp - damageTaken ) / Monster::GetHitPoints();
+    return unitsLeft * ( Monster::GetDamageMin() + Monster::GetDamageMax() ) / 2;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_troop.cpp:484]: This is incorrect. We don't use Bless and Curse modificators here.

**Review by ihhub — COMMENTED** [review]
Hi @idshibanov , I left just few comments. Please take a look.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 2 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:54]: "Faster" is ambiguous here. This call determines how far unit can move in context of threat range. Initiative/turn order aspect of the speed wouldn't matter.

File: src/fheroes2/battle/battle_troop.cpp (hunk 5 above)
Comment (idshibanov) [reply src/fheroes2/battle/battle_troop.cpp:484]: Good catch.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 6]
```diff
@@ -131,70 +211,139 @@ namespace AI
...
-                // 2. Calculate enemy retaliation after
-                // 3. Update damageDiff if it's bigger than current value
-                // 4. Save target selection
+                // force archer to fight back by setting initial expectation to lowest possible (if we're losing battle)
+                int bestOutcome = ( myArmyStrength < enemyArmyStrength ) ? -highestDamageExpected : 0;
+                bool canOutrunEnemy = true;
+
+                const Indexes & adjacentEnemies = Board::GetAdjacentEnemies( currentUnit );
+                for ( const int cell : adjacentEnemies ) {
+                    const Unit * enemy = Board::GetCell( cell )->GetUnit();
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:231]: Are we sure that `enemy` is always not NULL?

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 7]
```diff
@@ -131,70 +211,139 @@ namespace AI
...
+                    const int archerMeleeDmg = currentUnit.GetDamage( *enemy );
+                    const int damageDiff = archerMeleeDmg - enemy->CalculateRetaliationDamage( archerMeleeDmg );
+
+                    if ( bestOutcome < damageDiff ) {
+                        bestOutcome = damageDiff;
+                        target = enemy;
+                        targetCell = cell;
+                    }
+                    if ( canOutrunEnemy && !CheckIfUnitIsFaster( currentUnit, *enemy ) )
+                        canOutrunEnemy = false;
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:241]: This variable is for every unit while the rest of variables are not. We need a little explanation about this case please.

File: src/fheroes2/battle/battle_pathfinding.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+                newIndex = fromCell + ARENAW - ( isOddRow ? 1 : 0 );
+                wideUnitOffset = 1;
+            }
+            break;
+        case Battle::LEFT:
+            if ( x > 0 )
+                newIndex = fromCell - 1;
+            break;
+        default:
+            break;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_pathfinding.cpp:70]: We can `return -1` from here. No need to check further.

File: src/fheroes2/battle/battle_pathfinding.cpp [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                    node._isOpen = false;
+                    node._from = headIdx;
+                    node._cost = 1;
+                }
+            }
+        }
+        else {
+            std::vector<int> nodesToExplore;
+
+            if ( unitHead ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_pathfinding.cpp:153]: We check `!unitHead` at line 124 so there is no need to make such check here.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @idshibanov , I left few comments. Could you please take a look into them?

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 6 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:235]: `GetAdjacentEnemies` must return valid tiles. I will log an error if it fails now.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 7 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp]: Added a comment there, hopefully it's enough. To stay true this condition has to pass for every adjacent enemy unit

File: src/fheroes2/ai/normal/ai_normal_battle.cpp [hunk 10]
```diff
@@ -34,30 +34,73 @@ using namespace Battle;
 
 namespace AI
 {
-    bool CheckBattleRetreat()
+    bool isHeroWorthSaving( const Heroes * hero )
     {
-        // FIXME: force AI retreat from battle since battle logic is not implemented
-        return true;
+        return hero && ( hero->GetLevel() > 2 || !hero->GetBagArtifacts().empty() );
```
Comment (ihhub) [inline src/fheroes2/ai/normal/ai_normal_battle.cpp:39]: Shall we change the conditions to check primary skills instead? For example, some Campaign hero's are definitely worth to be saved as they have much higher stats.

File: src/fheroes2/battle/battle_arena.cpp [hunk 11]
```diff
@@ -564,6 +567,27 @@ Battle::Indexes Battle::Arena::GetPath( const Unit & b, const Position & dst )
...
+    //_pathfinder.calculate( unit.GetPosition(), unit.isWide() );
+    return _pathfinder.getPath( indexTo );
+}
+
+uint32_t Battle::Arena::CalculateMoveDistance( int32_t indexTo )
+{
+    return _pathfinder.getDistance( indexTo );
+}
+
+bool Battle::Arena::TileIsAccessible( int32_t indexTo )
```
Comment (ihhub) [inline src/fheroes2/battle/battle_arena.cpp:581]: Battle grid contains hexes, not tiles. Maybe it's worth to rename to don't make a confusion for other developers.

File: src/fheroes2/battle/battle_troop.cpp [hunk 12]
```diff
@@ -474,6 +474,25 @@ u32 Battle::Unit::GetSpeed( bool skip_standing_check ) const
     return speed;
 }
 
+uint32_t Battle::Unit::CalculateRetaliationDamage( uint32_t damageTaken ) const
```
Comment (ihhub) [inline src/fheroes2/battle/battle_troop.cpp:477]: Please check `u32 Battle::Unit::GetDamage( const Unit & enemy ) const` method and maybe remove duplicated code.

Comment (ihhub) [reply src/fheroes2/battle/battle_troop.cpp:477]: At least
```
    uint32_t damagePerUnit = 0;
    if ( Modes( SP_CURSE ) )
        damagePerUnit = Monster::GetDamageMin();
    else if ( Modes( SP_BLESS ) )
        damagePerUnit = Monster::GetDamageMax();
    else
        damagePerUnit = ( Monster::GetDamageMin() + Monster::GetDamageMax() ) / 2;
```
we can definitely wrap into a separate code.

**Review by ihhub — COMMENTED** [review]
Hi @idshibanov , good work! I left just few comments here. Please take a look.

File: src/fheroes2/ai/normal/ai_normal_battle.cpp (hunk 10 above)
Comment (idshibanov) [reply src/fheroes2/ai/normal/ai_normal_battle.cpp:39]: We could change it, but I think it's not worth the effort at the moment. Ideally the threshold/condition will be calculated during the AI turn, before the battle.

File: src/fheroes2/battle/battle_troop.cpp (hunk 12 above)
Comment (idshibanov) [reply src/fheroes2/battle/battle_troop.cpp:477]: Code looks similar but it achieves different things. Retaliation calculates raw damage (from `Monster` class) numbers without automatically applying unit count (in `ArmyTroop` class) and damage modifiers.

I think a lot of that code (`GetDamage`) has to be refactored anyway, I just didn't want to turn it into another chunk of changes.

**Comment by ihhub** [comment]
Hi @idshibanov thanks a lot for these changes!

**Pull request merged**